koku/masu/external/downloader/gcp/gcp_report_downloader.py

```diff
@@ -0,0 +1,316 @@
+"""GCP Report Downloader."""
+
+import logging
+import os
+
+from dateutil.relativedelta import relativedelta
+from dateutil.rrule import DAILY, rrule
+from google.cloud import storage
+from providers.gcp.provider import GCPProvider
+from rest_framework.exceptions import ValidationError
+
+from masu.config import Config
+from masu.external import UNCOMPRESSED
+from masu.external.downloader.downloader_interface import DownloaderInterface
+from masu.external.downloader.report_downloader_base import ReportDownloaderBase
+
+
+DATA_DIR = Config.TMP_DIR
+LOG = logging.getLogger(__name__)
+
+
+class GCPReportDownloaderError(Exception):
+    """GCP Report Downloader error."""
+
+
+class GCPReportDownloaderNoFileError(Exception):
+    """GCP Report Downloader error for missing file."""
+
+
+class GCPReportDownloader(ReportDownloaderBase, DownloaderInterface):
+    """
+    GCP Cost and Usage Report Downloader.
+
+    For configuration of GCP, see
+    https://cloud.google.com/billing/docs/how-to/export-data-file
+    """
+
+    def __init__(self, customer_name, bucket_name, **kwargs):
+        """
+        Constructor.
+
+        Args:
+            customer_name (str): Name of the customer
+            bucket_name (str): Name of the GCP bucket containing the CUR
+
+        """
+        super().__init__(**kwargs)
+
+        self.bucket_name = bucket_name
+        self.customer_name = customer_name.replace(' ', '_')
+        self._provider_id = kwargs.get('provider_id')
+
+        try:
+            GCPProvider().cost_usage_source_is_reachable(None, bucket_name)
+            self._storage_client = storage.Client()
+            self._bucket_info = self._storage_client.lookup_bucket(self.bucket_name)
+        except ValidationError as ex:
+            LOG.error(
+                'GCP bucket %(bucket_name)s for customer %(customer_name)s is not reachable. '
+                'Error: %(ex)s',
+                {
+                    'customer_name': customer_name,
+                    'bucket_name': bucket_name,
+                    'ex': str(ex),
+                },
+            )
+            raise GCPReportDownloaderError(str(ex))
+
+    def get_report_context_for_date(self, date_time):
+        """
+        Get the report context for a provided date.
+
+        Args:
+            date_time (datetime.datetime): The starting datetime object
+
+        Returns:
+            ({}) Dictionary containing the following keys:
+                manifest_id - (String): Manifest ID for ReportManifestDBAccessor
+                assembly_id - (String): unique ID for this provider and date range
+                compression - (String): Report compression format
+                files       - ([]): List of report files.
+
+        """
+        manifest_data = self._generate_monthly_pseudo_manifest(date_time)
+
+        if not self.check_if_manifest_should_be_downloaded(
```

Comment:
I'm not convinced this will download files added in the same month. 

Say you add the provider on Oct 15th. The stored manifest will have something like:
```
{
    'assembly_id': '1:2019-10-1:2019-10-31',
    'billing_period_start_datetime': '2019-10-1',
    'num_total_files': 15,
    'num_processed_files': 15
}
```

On the next run on 2019-10-16 `check_if_manifest_should_be_downloaded` will return `False`, because the manifest exists and `num_processed_files < num_total_files` on the stored manifest. Even though the new pseudo manifest data should have 16 files.

Let me know if I missed something in some parent class that makes sure this doesn't happen.

Reply:
I changed the way the assembly ID is generated to include the file count. So, when new files appear, the number will increment, a different assembly ID will be generated, and `check_if_manifest_should_be_downloaded` should indicate that the manifest needs to download and process.

This does mean we will be needlessly downloading the same previous days' files and reprocessing them, but this compromise is simpler and quicker than redesigning how these manifest-y time periods are handled at the higher level across all provider types (which I would prefer, but is a much larger and riskier change).